Furtherscape

sheavalentine-mdsol opened the pull request:
This is basic functionality allowing making requests that are unsafe or require a body such as DELETE, PUT, etc...

sheavalentine-mdsol committed:
added basic stubs for parameterized calls

sheavalentine-mdsol committed:
crud tests

Commit:
uri => url

Commit:
remove unecessary return

Commit:
refactor to remove debugging, condense code

Commit:
add pry require for (my) sanity

Commit:
add create spec

Commit:
add delete draft scenario

Commit:
fix delete spec, add todo for crichton test service

Commit:
make delete spec pass

Commit:
rename resource crud to resource crud spec

CGDS commented on spec/lib/farscape/integration/resource_crud_spec.rb:
```diff
@@ -0,0 +1,55 @@
+      let(:name) { 'Brave New DRD' }
+      let(:params) { {name: name}.merge(can_do_hash) }
+
+      #TODO fix crichton test service, its printing out the entire collection of DRDs.  Also, add
```
This TODO no longer applies.  And its wrong.  Woops.

CGDS commented on spec/lib/farscape/integration/resource_crud_spec.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+  end
+
+  # start at the entry entry_point
```
Can we remove this comment?  Still need toggle and update tests though.

CGDS commented on lib/farscape/transition.rb:
```diff
@@ -1,16 +1,29 @@
 module Farscape
   class Transition < Representors::Transition
 
+    EMPTY_BODIES = { hale: "{}" }
```
Did we agree on a TODO to fix this behavior in representors?

csavage-mdsol commented on spec/lib/farscape/integration/resource_crud_spec.rb:
```diff
@@ -0,0 +1,55 @@
+      expect(drds_resource.transitions.keys).to include('create')
+    end
+
+    it 'can create a drd' do
```
Oh, we should blow up this drd at the end of the test.  I think we test total count somewhere.

sheavalentine-mdsol replied:
?

HonoreDB replied:
I assume this led to conversation offline? Connor seems right to me, this looks like it leaks state.

sheavalentine-mdsol replied:
It is a state leak into things that should be robust against a change in that state.  
Adding a delete confuses what the code is testing, and adding an "after do" can confuse the actual source of the error.

HonoreDB replied:
So you want tests to avoid assuming that no drds exist at the start of the test? That's reasonable, but it's not how we normally do things so we might want to put a warning somewhere.

sheavalentine-mdsol replied:
Suggestions?

HonoreDB replied:
My first preference would be a high-level (like in spec_helper) after hook, so if the iterate-over-all-drds-and-delete-them  code fails it'll cause all specs to fail and the cause will be obvious. If you don't want to do that I'd settle for a note in CONTRIBUTING.md (under a new Testing section).

sheavalentine-mdsol replied:
I went with the simplest solution.

sheavalentine-mdsol committed:
changes as per PR

jcarres-mdsol commented on lib/farscape/transition.rb:
```diff
@@ -1,16 +1,29 @@
 module Farscape
   class Transition < Representors::Transition
 
+    EMPTY_BODIES = { hale: "{}" } #TODO: Fix Representor to allow nil resources
```
why hack it? fix them instead

sheavalentine-mdsol replied:
Since Representors is a different package it's out of this PR's scope to fix it.

jcarres-mdsol commented on spec/lib/farscape/integration/resource_crud_spec.rb:
```diff
@@ -0,0 +1,51 @@
+        expect(error_attributes).to_not eq(self_attributes)
+      end
+      
+      xit 'can update a drd' do
```
either implement them or not add them

HonoreDB replied:
Why not, isn't this what pending specs are for?

jcarres-mdsol replied:
What's the point of comiting empty pending specs? I do not think it adds value

HonoreDB replied:
Ah, you're objecting to having the spec body be empty. I think the description still serves the purpose of speccing the pending feature, though. It's just a TODO.

sheavalentine-mdsol replied:
Yeah, it's a TODO.  I originally had it as a comment, but felt that a pending spec was better.  Any opinions of RSG on this?

miwest929 commented on lib/farscape/transition.rb:
```diff
@@ -1,16 +1,29 @@
 require 'representors'
+require 'ostruct'
```
According to this stackoverflow post: http://stackoverflow.com/questions/1177594/ruby-struct-vs-openstruct `OpenStruct` is much, much slower than `Struct`. Is that a concern? Does the performance loss outweigh the enhanced expressiveness?

sheavalentine-mdsol replied:
Nothing complicated is being done with OpenStruct, though in the future it should probably be replaced with an object that has specific setters that are supported.
Should I add a TODO?

sheavalentine-mdsol commented:
So, this can't be merged until crichton_test_service is a proper MDSOL gem right?

HonoreDB commented:
I don't see a Gemfile change in the diff.

sheavalentine-mdsol commented:
Yeah, but I don't think brave_new_develop can be merged either for the same reason.

```
Permission denied (publickey).
fatal: Could not read from remote repository.
Please make sure you have the correct access rights
and the repository exists.
Git error: command `git clone
'git@github.com:csavage-mdsol/crichton_test_service.git'
"/home/travis/build/mdsol/farscape/vendor/bundle/ruby/1.9.1/cache/bundler/git/crichton_test_service-5ed308d2ebb633303c8fa9dad4756bc1904db78e"
--bare --no-hardlinks --quiet` in directory /home/travis/build/mdsol/farscape
has failed.
```

HonoreDB commented:
Right.

csavage-mdsol commented:
We're working on getting Crichton Test Service pulled into an mdsol repo.  Should happen today.

sheavalentine-mdsol committed:
cleaned up after create

I decided to do this the simple way with a comment that should show up in the event of a misplaced failure.

It seemed easier than the other solutions

sheavalentine-mdsol committed:
rest of crud spec

sheavalentine-mdsol committed:
do blocks for multiline

sheavalentine-mdsol committed:
Implement Contentious Stipped Down Interface

This, to the best of my ability, ensures the interface as desgined in the documentation.
However there are several barriers to completely matching the documented interface.

1. Moya needs to be fleshed out more and support more.
2. Several documented use cases were inconsistent.
3. An Exact Error set and use cases need to be defined
4. The Interface does not consider field objects - see below.
5. Empty response bodies does not have well specified behavior
6. Moya needs to also serve a Leviathan service to show resource bridging

Regarding Fields:
The design called for
```
resource.attributes[field] # => 'field value'
```
This interface makes it impossible to discover important information about the field like it's scope, it's type, or its doc.

So instead you have to do
```
resource.attributes[field].value # => 'field value'
```
but this is ugly, this is why I proposed

```
resource.attributes[field] # => Representor::FieldObject
resource[field] # => field.value
```

Perhaps this is not the best interface, but the documented set doesn't cover all usecases.

sheavalentine-mdsol committed:
White space changes
added some more TODO's
Changes as per PR

miwest929 commented:
@HonoreDB @csavage-mdsol What's blocking this PR from being merged?

miwest929 commented:
@sheavalentine-mdsol After speaking with @csavage-mdsol we should use `moya` develop as it has same functionality as `crichton-test-service`. This should allow the Travis CI build to pass. @csavage-mdsol Anything to add?

sheavalentine-mdsol committed:
repointed to moya

csavage-mdsol commented:
Nope, that should do it.

sheavalentine-mdsol committed:
https used for gem

sheavalentine-mdsol committed:
https for crichton too

sheavalentine-mdsol committed:
crichton https

sheavalentine-mdsol committed:
okay mdsol-share...

sheavalentine-mdsol committed:
byebug

Commit:
make byebug conditional on ruby version

miwest929 commented:
@sheavalentine-mdsol @csavage-mdsol Let's make same changes here as we did on the other farscape PR so that the Travis build can pass.

sheavalentine-mdsol commented:
As soon as the other PR gets merged it will automatically make this one mergable, because it's targeting this one.

Commit:
Merge pull request #19 from mdsol/feature/interface_spec_implement

Implement Contentious Stipped Down Interface

miwest929 commented:
@csavage-mdsol @HonoreDB Anymore comments?

csavage-mdsol commented:
@miwest929 nope

miwest929 commented:
merging...
